File: crates/gjc-config/src/env_file.rs
//! Environment-file parsing primitives.
//!
//! Port of `packages/utils/src/env-file.ts` + the relevant checks from
//! `packages/utils/src/spawn-env.ts`. Parsing is intentionally identical:
//! dotenv keys must be POSIX shell identifiers, values must not contain NUL,
//! and shell startup files are parsed without executing user shell code.

use std::collections::HashMap;
use std::path::Path;

/// Strict shell-identifier shape: `[A-Za-z_][A-Za-z0-9_]*`.
pub fn is_valid_env_name(name: &str) -> bool {
	let mut chars = name.chars();
	match chars.next() {
		Some(c) if c.is_ascii_alphabetic() || c == '_' => {},
		_ => return false,
	}
	chars.all(|c| c.is_ascii_alphanumeric() || c == '_')
}

/// Values must not contain NUL (mirrors `isSafeEnvValue`).
pub fn is_safe_env_value(value: &str) -> bool {
	!value.contains('\0')
}

/// Strip an unquoted ` #comment` suffix, honoring `\` escapes and quotes.
fn strip_inline_shell_comment(value: &str) -> &str {
	let bytes: Vec<char> = value.chars().collect();
	let mut quote: Option<char> = None;
	let mut i = 0;
	let mut byte_pos = 0;
	while i < bytes.len() {
		let c = bytes[i];
		if c == '\\' {
			byte_pos += c.len_utf8() + bytes.get(i + 1).map_or(0, |n| n.len_utf8());
			i += 2;
			continue;
		}
		if (c == '"' || c == '\'') && (quote.is_none() || quote == Some(c)) {
			quote = if quote.is_some() { None } else { Some(c) };
			byte_pos += c.len_utf8();
			i += 1;
			continue;
		}
		if c == '#' && quote.is_none() && (i == 0 || bytes[i - 1].is_whitespace()) {
			return value[..byte_pos].trim_end();
		}
		byte_pos += c.len_utf8();
		i += 1;
	}
	value.trim_end()
}

fn unquote(value: &str) -> &str {
	if value.len() >= 2
		&& ((value.starts_with('"') && value.ends_with('"'))
			|| (value.starts_with('\'') && value.ends_with('\'')))
	{
		&value[1..value.len() - 1]
	} else {
		value
	}
}

/// Parse a `.env` file. Missing/unreadable files yield an empty map,
/// mirroring `parseEnvFile`.
pub fn parse_env_file(path: &Path) -> HashMap<String, String> {
	let mut result = HashMap::new();
	let Ok(content) = std::fs::read_to_string(path) else {
		return result;
	};
	for line in content.split('\n') {
		let trimmed = line.trim();
		if trimmed.is_empty() || trimmed.starts_with('#') {
			continue;
		}
		let Some(eq) = trimmed.find('=') else {
			continue;
		};
		let key = trimmed[..eq].trim();
		if !is_valid_env_name(key) {
			continue;
		}
		let value = unquote(trimmed[eq + 1..].trim());
		if !is_safe_env_value(value) {
			continue;
		}
		result.insert(key.to_owned(), value.to_owned());
	}
	result
}

/// Parse simple POSIX shell assignments without executing shell code.
///
/// Handles `export KEY=value` / `KEY=value` in startup files like `~/.zshrc`.
/// Mirrors `parseShellEnvFile`: dynamic values containing `$` or backticks
/// are ignored.
pub fn parse_shell_env_file(path: &Path) -> HashMap<String, String> {
	let mut result = HashMap::new();
	let Ok(content) = std::fs::read_to_string(path) else {
		return result;
	};
	for line in content.split('\n') {
		let trimmed = line.trim();
		if trimmed.is_empty() || trimmed.starts_with('#') {
			continue;
		}
		let rest = trimmed
			.strip_prefix("export ")
			.map_or(trimmed, str::trim_start);
		let Some(eq) = rest.find('=') else { continue };
		let key = rest[..eq].trim_end();
		if !is_valid_env_name(key) {
			continue;
		}
		let mut value = strip_inline_shell_comment(&rest[eq + 1..])
			.trim()
			.to_owned();
		if let Some(stripped) = value.strip_suffix(';') {
			value = stripped.trim_end().to_owned();
		}
		let value = unquote(&value);
		if !is_safe_env_value(value) || value.contains('$') || value.contains('`') {
			continue;
		}
		result.insert(key.to_owned(), value.to_owned());
	}
	result
}

#[cfg(test)]
mod tests {
	use super::*;
	use std::io::Write;

	fn write_temp(content: &str) -> tempfile::NamedTempFile {
		let mut f = tempfile::NamedTempFile::new().unwrap();
		f.write_all(content.as_bytes()).unwrap();
		f
	}

	#[test]
	fn env_file_basic() {
		let f = write_temp("FOO=bar\n# comment\nBAZ='quoted'\nQUX=\"dq\"\nBAD KEY=x\n1BAD=y\n");
		let map = parse_env_file(f.path());
		assert_eq!(map.get("FOO").unwrap(), "bar");
		assert_eq!(map.get("BAZ").unwrap(), "quoted");
		assert_eq!(map.get("QUX").unwrap(), "dq");
		assert!(!map.contains_key("BAD KEY"));
		assert!(!map.contains_key("1BAD"));
	}

	#[test]
	fn env_file_missing_returns_empty() {
		assert!(parse_env_file(Path::new("/nonexistent/.env")).is_empty());
	}

	#[test]
	fn shell_env_file_rules() {
		let f = write_temp(concat!(
			"export A=1\n",
			"B=two # trailing comment\n",
			"C=\"has # inside\"\n",
			"D=$HOME/dynamic\n",
			"E=`cmd`\n",
			"F='literal';\n",
		));
		let map = parse_shell_env_file(f.path());
		assert_eq!(map.get("A").unwrap(), "1");
		assert_eq!(map.get("B").unwrap(), "two");
		assert_eq!(map.get("C").unwrap(), "has # inside");
		assert!(!map.contains_key("D"));
		assert!(!map.contains_key("E"));
		assert_eq!(map.get("F").unwrap(), "literal");
	}
}

File: packages/utils/src/env-file.ts
/**
 * Environment-file parsing primitives.
 *
 * Kept in a leaf module so both `env.ts` and `dirs.ts` can use them. `env.ts`
 * imports `dirs.ts`, so anything `dirs.ts` needs from the env layer has to live
 * below both of them.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { isSafeEnvValue } from "./spawn-env";

const ENV_NAME_RE = /^[A-Za-z_][A-Za-z0-9_]*$/;

/**
 * Strict shell-identifier shape. Used for dotenv keys we accept into
 * `Bun.env` — those should be referenceable as `$NAME` from POSIX shells,
 * so we reject anything outside `[A-Za-z_][A-Za-z0-9_]*`.
 */
export function isValidEnvName(name: string): boolean {
	return ENV_NAME_RE.test(name);
}

function stripInlineShellComment(value: string): string {
	let quote: '"' | "'" | undefined;
	for (let i = 0; i < value.length; i++) {
		const char = value[i];
		if (char === "\\") {
			i++;
			continue;
		}
		if ((char === '"' || char === "'") && (!quote || quote === char)) {
			quote = quote ? undefined : char;
			continue;
		}
		if (char === "#" && !quote && (i === 0 || /\s/.test(value[i - 1] ?? ""))) {
			return value.slice(0, i).trimEnd();
		}
	}
	return value.trimEnd();
}

/**
 * Strips an unquoted trailing `# comment` from a dotenv value the way Bun's
 * dotenv loader does: an unescaped `#` starts a comment regardless of the
 * preceding character (`a#b` loads as `a`), while `#` inside quotes or after a
 * backslash escape survives. Used only by `parseEnvFile`; shell files use
 * `stripInlineShellComment`, whose POSIX rule requires whitespace before `#`.
 */
function stripInlineDotenvComment(value: string): string {
	let quote: '"' | "'" | undefined;
	for (let i = 0; i < value.length; i++) {
		const char = value[i];
		if (char === "\\") {
			i++;
			continue;
		}
		if ((char === '"' || char === "'") && (!quote || quote === char)) {
			quote = quote ? undefined : char;
			continue;
		}
		if (char === "#" && !quote) return value.slice(0, i).trimEnd();
	}
	return value.trimEnd();
}

/**
 * Parses simple POSIX shell environment assignments from files such as
 * ~/.zshrc without executing user shell code. Supports `export KEY=value` and
 * `KEY=value`, including single/double quoted literal values. Dynamic shell
 * expressions are intentionally ignored because evaluating startup files would
 * run arbitrary code during CLI startup.
 */
export function parseShellEnvFile(filePath: string): Record<string, string> {
	const result: Record<string, string> = {};
	try {
		const content = fs.readFileSync(filePath, "utf-8");
		for (const line of content.split("\n")) {
			const trimmed = line.trim();
			if (!trimmed || trimmed.startsWith("#")) continue;

			const match = /^(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)=(.*)$/.exec(trimmed);
			if (!match) continue;

			const key = match[1];
			if (!isValidEnvName(key)) continue;

			let value = stripInlineShellComment(match[2] ?? "").trim();
			if (value.endsWith(";")) value = value.slice(0, -1).trimEnd();
			if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
				value = value.slice(1, -1);
			}
			if (!isSafeEnvValue(value)) continue;
			if (/[$`]/.test(value)) continue;

			result[key] = value;
		}
	} catch {
		// File doesn't exist or can't be read - return empty result
	}

	return result;
}

/**
 * Parses a .env file synchronously and extracts key-value string pairs.
 * Ignores lines that are empty or start with '#'. Trims whitespace.
 * Allows values to be quoted with single or double quotes.
 * Returns an object of key-value pairs.
 *
 * The trust guards (`trustedAgentDirOverrideFor`, `trustedConfigDirName`,
 * `filterCredentialInheritedEnv`) decide provenance by comparing
 * `process.env` against this parse, so the accepted syntax must be a superset
 * of what Bun's own dotenv loader honors in `cwd/.env`: `export KEY=value`,
 * whitespace around `=` or `:`, and `#` comments after unquoted values (quotes keep
 * their `#`). Values that Bun would expand (`$VAR`, `${VAR}`, backticks,
 * command substitution) are kept as their literal text: the trust rule only
 * needs the parser to see the key at all, and an operator environment value
 * cannot equal attacker-written expansion text, so a literal parse stays
 * conservative.
 */
export function parseEnvFile(filePath: string): Record<string, string> {
	try {
		return parseEnvFileContent(fs.readFileSync(filePath, "utf-8"));
	} catch {
		// File doesn't exist or can't be read - return empty result
		return {};
	}
}

/** Parse dotenv content that has already been read from a trusted file. */
export function parseEnvFileContent(content: string): Record<string, string> {
	const result: Record<string, string> = {};
	for (const line of content.split("\n")) {
		const trimmed = line.trim();
		// Skip comments and blank lines
		if (!trimmed || trimmed.startsWith("#")) continue;

		const match = /^(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*(?:=|:)\s*(.*)$/.exec(trimmed);
		if (!match) continue;

		const key = match[1];
		if (!isValidEnvName(key)) continue;

		// Strip an unquoted trailing `# comment` the way Bun's dotenv loader
		// does (`KEY=v#note` loads as `v`); quoted `#` survives.
		let value = stripInlineDotenvComment(match[2] ?? "").trim();

		// Remove surrounding quotes (" or ')
		if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
			value = value.slice(1, -1);
		}
		if (!isSafeEnvValue(value)) continue;

		result[key] = value;
	}

	return result;
}

/**
 * What the caller's checkout declares through its dotenv files.
 *
 * `values` is the merged declaration set, later layers winning. `dynamic` holds
 * the keys whose surviving declaration is one Bun expands at load time, which
 * every provenance guard refuses outright because a value comparison cannot see
 * what such a declaration became.
 */
export interface ProjectEnvSnapshot {
	values: Record<string, string>;
	dynamic: Set<string>;
}

/**
 * Windows environment variable names are case-insensitive, so a project dotenv
 * line `userprofile=...` is what `process.env.USERPROFILE` resolves to. Every
 * provenance lookup is spelled in upper case, so the snapshot must be keyed the
 * same way or the declaration is invisible to the guard while still being live
 * in the process. POSIX names are case-sensitive and must not fold.
 */
export function canonicalEnvKey(name: string): string {
	return process.platform === "win32" ? name.toUpperCase() : name;
}

/**
 * The layered dotenv declarations Bun overlays into `process.env` for a cwd.
 *
 * Lives in this leaf module so every consumer shares ONE notion of provenance.
 * `dirs.ts` resolves the config, agent and log directories from it, and
 * `scripts/test-preload.ts` decides test isolation from it — a second, narrower
 * reader (only `cwd/.env`, its own regex, its own dynamic test) is how a
 * `GJC_LOG_DIR` declared in `.env.local` / `.env.$NODE_ENV` came to be honored
 * by the preload and then rejected in production, silently routing test log
 * records to the operator's canonical sink. This module imports only `node:fs`,
 * `node:path` and the import-free `./spawn-env`, so importing it has no side
 * effects and cannot freeze resolver state the way importing `dirs.ts` would.
 *
 * `.env.local` is deliberately skipped when `NODE_ENV === "test"`, matching the
 * convention that a local override file is not part of a test run.
 */
export function projectEnvSnapshot(cwd = process.cwd()): ProjectEnvSnapshot {
	const nodeEnv = process.env.NODE_ENV;
	const validNodeEnv = nodeEnv && /^[A-Za-z0-9_-]+$/.test(nodeEnv) ? nodeEnv : undefined;
	const files = [
		".env",
		...(validNodeEnv ? [`.env.${validNodeEnv}`] : []),
		...(validNodeEnv !== "test" ? [".env.local"] : []),
		...(validNodeEnv ? [`.env.${validNodeEnv}.local`] : []),
	];
	const values: Record<string, string> = {};
	const dynamic = new Set<string>();
	for (const file of files) {
		for (const [rawKey, value] of Object.entries(parseEnvFile(path.join(cwd, file)))) {
			const key = canonicalEnvKey(rawKey);
			values[key] = value;
			if (/[$`]/.test(value)) dynamic.add(key);
			else dynamic.delete(key);
		}
	}
	return { values, dynamic };
}
